File: Completed/CSharp/20.cs
using System.Collections;
using System.Linq;

public class Solution {  
    public bool checkMatch(char x, char y) {
        bool valid = false;
        switch(x) {
            case '(':
                valid = y == ')';
                break;
            case '[':
                valid = y == ']';
                break;
            case '{':
                valid = y == '}';
                break;
            case ')':
                valid = y == '(';
                break;
            case ']':
                valid = y == '[';
                break;
            case '}':
                valid = y == '{';
                break;
            default:
                valid = false;
                break;
        }
        return valid;
    }
    
    public bool IsValid(string s) {
        var charStk = new Stack<char>();
        
        if(s.Length % 2 != 0) {
            return false;
        }
        
        foreach(var c in s) {
            if(charStk.Count() > 0 && checkMatch(c,charStk.Peek())){
                charStk.Pop();
            }
            else {
                charStk.Push(c);
            }
            
        }
        
        return charStk.Count() == 0;
    }
}

File: Completed/Go/20.go
type runeStack struct {
	data []rune
}

func isValid(s string) bool {
	stk := runeStack{}
	for _, r := range s {
		if top := stk.Peek(); isPair(top, r) {
			stk.Pop()
		} else {
			stk.Push(r)
		}
	}

	return stk.Len() == 0
}

func (this *runeStack) Peek() rune {
	var top rune
	if len(this.data) != 0 {
		top = this.data[len(this.data)-1]
	}
	return top
}

func (this *runeStack) Push(x rune) {
	this.data = append(this.data, x)
}

func (this *runeStack) Pop() rune {
	x := this.Peek()
	this.data = this.data[:len(this.data)-1]

	return x
}

func (this *runeStack) Len() int {
	return len(this.data)
}

func isPair(a, b rune) bool {
	m := map[rune]rune{
		'(': ')', '[': ']', '{': '}',
	}
	val, ok := m[a]
	return ok && val == b
}